Add tensor blog post (#114)
vespa-engine/blog: _posts/2021-09-30-computing-with-tensors.md

---
layout: post
title: "Computing with tensors in Vespa"
date: '2021-09-30'
tags: []
author: lesters
image: assets/2021-09-30-computing-with-tensors/pietro-jeng-n6B49lTx7NM-unsplash.jpg
skipimage: true

excerpt: "In this blog post, we'll explore some of the unique properties of tensors in Vespa."
---

![Decorative image](/assets/2021-09-30-computing-with-tensors/pietro-jeng-n6B49lTx7NM-unsplash.jpg)
<p class="image-credit">
Photo by <a href="https://unsplash.com/@pietrozj?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Pietro Jengr</a> on <a href="https://unsplash.com/photos/n6B49lTx7NM?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText">Unsplash</a>
</p>

In computer science, a tensor is a data structure that generalizes scalars,
vectors, matrices, and higher-order structures into a single construct.

In Vespa we have introduced a tensor formalism that differs from what one
usually finds in most modern machine learning frameworks today. The main
differences are:

- Named dimensions
- Unified sparse and dense tensor types
- A small but very powerful set of core functions

> In this blog post, we'll explore the differences and the benefits of tensors in
> Vespa. We will also introduce the recently released [tensor

Reviewer: compared to what? maybe just simplify:
```suggestion
In this blog post, we'll explore different aspects of tensors in Vespa.
We will also introduce the [tensor
```

> playground](https://docs.vespa.ai/playground/), a tool to get familiar with and

Reviewer: just realized, we should have a Vespa logo on the playground site, too (unrelated to this post, though)

Author: Good idea!

challenging.

However, the tensor does not contain the information itself to help describe the
dimensions. As a result, practitioners often use comments such as the following
to document the dimension order:

```
# Num x Height x Width x Channel
tensor = torch.tensor(numpy.load("images.npy"))
tensor.shape
> torch.Size([100, 64, 64, 3])
```

This is obviously error-prone.

In Vespa, we have taken a different approach and introduced named dimensions.
This enables a strong tensor type system. An example of a tensor type in Vespa
which holds the same data as above:

```
tensor(num[100], height[64], width[64], channel[3])
```

This type system provides more formal documentation, which makes it easier to
work with for humans. It also introduces the ability for tensors and tensor
operations to be semantically verified. This means we can perform static type
inference for all computation, catching errors at an early compile-time stage
rather than during runtime.

Later in the post, we'll see how this enables arbitrarily complex computation
with only a minimal, concise set of core operations.

## Sparse and dense tensors

> The tensor as a multi-dimensional array is often considered to be dense. This

Reviewer: ```suggestion
The tensor as a multi-dimensional array is often considered to be _dense_. This
```

means that all combinations of dimension indexes have a value, even if that
value is `0` or `NaN`. However, a tensor with many such values could be

> represented more efficiently as a sparse tensor, where only the non-empty values

Reviewer: ```suggestion
represented more efficiently as a _sparse_ tensor, where only the non-empty values
```

are defined. This can lead to considerable savings in space.

Unfortunately, the internal representation of sparse tensors in most frameworks
makes them incompatible with regular dense tensors. This leads to an entirely
separate set of functions operating on sparse and dense tensors, with functions
to convert between the two.

Vespa supports dense, sparse, and tensors that contain both dense and sparse

> dimensions, called a mixed tensor. A dense tensor is a tensor containing only

Reviewer: ```suggestion
dimensions, called a _mixed_ tensor. A dense tensor is a tensor containing only
```

- `join` - the natural join between two input tensors with a custom expression applied
- `merge` - merge two input tensors with a custom expression
- `concat` - concatenate two tensors along a given dimension

To aid developers, Vespa additionally provides higher-level, non-core
operations, which are all implemented in terms of these core operations.

This approach enables interoperability as implementing this small set is all
that is needed to realize complete support for tensor computation. Furthermore,
it makes optimization work more efficient. This is because low-level
optimizations are only required on this set of functions. Higher-level
optimizations can work on whichever chunks of these operations are beneficial,
independent of any chunking into higher-level functions humans happen to find
meaningful.

# The tensor formalism

In this section, we'll describe Vespa's full tensor formalism, including
tensors and tensor types, and the core set of tensor operations.

## Tensor types

A *tensor type* consists of a *value type* and a set of *dimensions*. The value type is a numeric data type: double, float, int8, bfloat16, etc. The default value type is double.

> A dimension consists of a *name* and a *type*. The type can be either *indexed* or *mapped*, and, if indexed, can optionally include a *size*. The number of dimensions in the tensor is called its *order*. A tensor without dimensions (0-order) is a single scalar value. A tensor with one dimension (first-order) is a vector. A tensor with two dimensions (second-order) is a matrix.

Reviewer: I learned something new then - I thought indexed tensors always had a size for each dimension. maybe add an example with an indexed tensor without size for a dimension, too

Author: This is only really supported on the Java side at the moment, but important for stateless model evaluation. So I'll think I'll let this be as it is for now.

Reviewer: all right!

### Vector inner product (dot product)

Again, given the two vectors `A` and `B` representing vectors:

```
A: tensor(x[3]):[1,2,3]
B: tensor(x[3]):[4,5,6]
```

Here, the vectors have the same dimension name. Given the exact same expression
as above, `A * B` or, equivalently, `join(A, B, f(a,b)(a*b))`, the result is
now:

```
tensor(x[3]):[4.0, 10.0, 18.0]
```

Since the two tensors now have a common dimension, the `join` matches the cells
with equal values in that dimension. So `x[0]` from `A` is combined with `x[0]`
from `B`, and so on. The lambda expression `f(a,b)(a * b)` defines these values
should be multiplied. This is called the element-wise product.

The actual inner (dot) product requires these values to be summed. Thus we can
add a `sum` operation so the expression becomes `sum(A * B)`. The `sum`
operation is actually a `reduce` operation, so this expands fully to the
following:

```

> reduce(join(A, B, f(a,b)(a,b)), sum)

Reviewer: ```suggestion
reduce(join(A, B, f(a,b)(a * b)), sum)
```

  multiplication, this is the same as an element-wise product.
- A `join` between tensors with disjunct dimensions is the cross join, or
  Cartesian product, between the input tensors.
- A `join` between tensors with partially overlapping dimensions produces
  results in-between these extremes, as seen in the matrix multiplication
  example above.

By combining `join` with `rename`, any such semantics can be achieved for any
tensors.

The `join` operation implements a natural join between tensors: it combines
cells in one tensor with cells from another based on the matching cells in their
common dimensions. It is thus the tensor language counterpart of the logical AND
operator.

## Summary

In this post, we've presented Vespa's tensor language. This formalism provides
generic computation over dense and sparse tensors, with full static type
inference, using just eight foundational tensor functions. Named dimensions are
used to add semantic information and achieve generality, and index labels
provide support for models working with strings.

In Vespa, this formalism makes numerical computation easier to express,
understand, and optimize.

> For more information, take a look at Vespas [tensor

Reviewer: ```suggestion
For more information, take a look at Vespa's [tensor
```
